node/core/approval-voting/src/criteria.rs

```diff
@@ -483,50 +619,100 @@ pub(crate) fn check_assignment_cert(
+			is_in_backing_group(&config.validator_groups, validator_index, *backing_group);
+
+		if is_in_backing {
+			return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		}
 	}
 
 	let &(ref vrf_output, ref vrf_proof) = &assignment.vrf;
-	match assignment.kind {
-		AssignmentCertKind::RelayVRFModulo { sample } => {
-			if sample >= config.relay_vrf_modulo_samples {
+	match &assignment.kind {
+		AssignmentCertKindV2::RelayVRFModuloCompact { sample, core_indices } => {
+			if *sample >= config.relay_vrf_modulo_samples {
 				return Err(InvalidAssignment(Reason::SampleOutOfBounds))
 			}
 
 			let (vrf_in_out, _) = public
 				.vrf_verify_extra(
-					relay_vrf_modulo_transcript(relay_vrf_story, sample),
+					relay_vrf_modulo_transcript(relay_vrf_story, *sample),
 					&vrf_output.0,
 					&vrf_proof.0,
-					assigned_core_transcript(claimed_core_index),
+					assigned_cores_transcript(core_indices),
 				)
 				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
 
+			let resulting_cores = relay_vrf_modulo_cores(&vrf_in_out, *sample + 1, config.n_cores);
+
+			// TODO: Enforce that all claimable cores are claimed, or include refused cores.
+			// Currently validators can opt out of checking specific cores.
+			// This is the same issue to how validator can opt out and not send their assignments in the first place.
+
+			// Ensure that the `vrf_in_out` actually includes all of the claimed cores.
+			if claimed_core_indices
+				.iter()
+				.fold(true, |cores_match, core| cores_match & resulting_cores.contains(core))
+			{
+				Ok(0)
+			} else {
+				gum::debug!(
+					target: LOG_TARGET,
+					?resulting_cores,
+					?claimed_core_indices,
+					"Assignment claimed cores mismatch",
+				);
+				Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+		},
+		AssignmentCertKindV2::RelayVRFModulo { sample } => {
+			if *sample >= config.relay_vrf_modulo_samples {
+				return Err(InvalidAssignment(Reason::SampleOutOfBounds))
+			}
+
+			let (vrf_in_out, _) = public
+				.vrf_verify_extra(
+					relay_vrf_modulo_transcript(relay_vrf_story, *sample),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_core_transcript(claimed_core_indices[0]),
+				)
+				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
+
+			let core = relay_vrf_modulo_core(&vrf_in_out, config.n_cores);
 			// ensure that the `vrf_in_out` actually gives us the claimed core.
-			if relay_vrf_modulo_core(&vrf_in_out, config.n_cores) == claimed_core_index {
+			if core == claimed_core_indices[0] {
 				Ok(0)
 			} else {
+				gum::debug!(
+					target: LOG_TARGET,
+					?core,
+					?claimed_core_indices,
+					"Assignment claimed cores mismatch",
+				);
 				Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
 			}
 		},
-		AssignmentCertKind::RelayVRFDelay { core_index } => {
-			if core_index != claimed_core_index {
+		AssignmentCertKindV2::RelayVRFDelay { core_index } => {
+			if *core_index != claimed_core_indices[0] {
```

Comment:
I'd think `claimed_core_indices` could be longer than one element here, so maybe check length or slice equality..
```
if [ *core_index ] != claimed_core_indices { ...
```

Reply:
That can happen only for the new assignments `RelayVRFModuloCompact`